backends/p4tools/common/lib/gen_eq.h

```diff
@@ -5,27 +5,23 @@
 
 namespace P4Tools {
 
-/// Generates an equality on a target expression and a keyset expression, recursing into lists.
+/// Generates an equality on two input expressions, recursing into lists and structs.
 /// This supports fuzzy matching on singleton lists: singleton lists are considered the same as
 /// their singleton elements. This is implemented by eagerly recursing into singleton lists before
 /// attempting to generate the equality.
 class GenEq {
  public:
-    static const IR::Expression *equate(const IR::Expression *target, const IR::Expression *keyset);
+    static const IR::Expression *equate(const IR::Expression *left, const IR::Expression *right);
 
  private:
-    static const IR::Expression *equate(const IR::Expression *target,
-                                        const IR::DefaultExpression *keyset);
-
-    static const IR::Expression *equate(const IR::Expression *target,
-                                        const IR::ListExpression *keyset);
-
-    static const IR::Expression *equate(const IR::Expression *target, const IR::Mask *keyset);
-
-    static const IR::Expression *equate(const IR::Expression *target, const IR::Range *keyset);
-
-    static const IR::Expression *equate(const IR::ListExpression *target,
-                                        const IR::ListExpression *keyset);
+    /// Recursively resolve lists of size 1 by returning the expression contained within.
+    static const IR::Expression *checkSingleton(const IR::Expression *expr);
+
+    /// Flatten and compare two lists.
+    /// Important, this equation assumes that struct expressions have been ordered.
+    /// This calculation does not match the names of the struct expressions.
```

Comment:
This is a little difficult to implement because it might need a lot of special casing. Let me think about something sensible. In general, the assumption is that the front-end has taking care of ordering.

Reply:
Actually, the right way to do this is to simply check the type of the underlying struct expression and match it up with the `components`. Unfortunately, that type is often a `Type_Name`, very inconvenient...

Reply:
Is it not possible to resolve the `Type_Name` using `P4::TypeMap` at this point?

Reply:
I have implemented this and it seems to pass all tests for now.

>Is it not possible to resolve the Type_Name using P4::TypeMap at this point?

We do not have access to the TypeMap at this point, P4Testgen does its own type resolution because passes can break type-checking.  Usually, any Type_Name is resolved before. The problem is Type_Names in nested struct expressions. We do not have to deal with this right now because we resolve most nesting.

Reply:
Now that https://github.com/p4lang/p4c/pull/4215 is merged, which resolves Type_Name, this is a much safer assumption and will hold.

Reply:
I see. Do you plan to add an additional check, or is this kept as a precondition only? I think this is actually complicated by the eager resolving of singletons which means the types don't need to match exactly. Therefore it is probably OK to just leave the "compatibility" of types as a precondition.